orchard_note_encryption.py

```diff
@@ -0,0 +1,295 @@
+        return bytes(cip.encrypt(b'\x00' * 12, plaintext))
+
+    @staticmethod
+    def decrypt(key, ciphertext):
+        cip = ChaCha20Poly1305(key)
+        return bytes(cip.decrypt(b'\x00' * 12, ciphertext))
+
+# https://zips.z.cash/protocol/nu5.pdf#saplingandorchardencrypt
+class OrchardNoteEncryption(object):
+    def __init__(self, random=os.urandom):
+        self._random = random
+
+    def rseed(self):
+        return self._random.b(32)
+
+    def encrypt(self, note: OrchardNote, memo, pk_d_new, g_d_new, cv_new, cm_new, ovk=None):
+        np = note.note_plaintext(memo)
+        esk = OrchardKeyAgreement.esk(np.rseed, note.rho)
+        p_enc = bytes(np)
+
+        epk = OrchardKeyAgreement.derive_public(esk, g_d_new)
+        ephemeral_key = bytes(epk)
+        shared_secret = OrchardKeyAgreement.agree(esk, pk_d_new)
+        k_enc = kdf_orchard(shared_secret, epk)
+        c_enc = OrchardSym.encrypt(k_enc, p_enc)
+
+        if not ovk:
+            ock = OrchardSym.k(self._random)
+            op = self._random.b(64)
+        else:
+            cv = bytes(cv_new)
+            cmx = bytes(cm_new.extract())
+            ock = prf_ock_orchard(ovk, cv, cmx, ephemeral_key)
+            op = bytes(pk_d_new) + bytes(esk)
+
+        c_out = OrchardSym.encrypt(ock, op)
+
+        self.esk = esk
+        self.shared_secret = shared_secret
+        self.k_enc = k_enc
+        self.p_enc = p_enc
+        self.ock = ock
+        self.op = op
+
+        return TransmittedNoteCipherText(
+            epk, c_enc, c_out
+        )
+
+class TransmittedNoteCipherText(object):
+    def __init__(self, epk, c_enc, c_out):
+        self.epk = epk
+        self.c_enc = c_enc
+        self.c_out = c_out
+
+    def decrypt_using_ivk(self, ivk: Scalar, rho, cm_star):
+        epk = self.epk
+        if not epk:
+            return None
+
+        shared_secret = OrchardKeyAgreement.agree(ivk, epk)
+        k_enc = kdf_orchard(shared_secret, epk)
+        p_enc = OrchardSym.decrypt(k_enc, self.c_enc)
+        if not p_enc:
+            return None
+
+        leadbyte = p_enc[0]
+        assert(leadbyte == 2)
+        np = OrchardNotePlaintext(
+            p_enc[1:12],   # d
+            Scalar.from_bytes(p_enc[12:20]),  # v
+            p_enc[20:52],  # rseed
+            p_enc[52:564], # memo
+        )
+
+        g_d = diversify_hash(np.d)
+        pk_d = OrchardKeyAgreement.derive_public(ivk, g_d)
+        note = OrchardNote(np.d, pk_d, np.v, rho, np.rseed)
+
+        esk = OrchardKeyAgreement.esk(np.rseed, rho)
+        if OrchardKeyAgreement.derive_public(esk, g_d) != epk:
+            return None
+
+        cm = note.note_commitment()
+        if not cm:
+            return None
+        if cm != cm_star:
+            return None
+
+        return (note, np.memo)
+
+    def decrypt_using_fvk(self, fvk, rseed, rho, cv, cm_star):
+        ock = prf_ock_orchard(fvk.ovk, bytes(cv), bytes(cm_star.extract()), bytes(self.epk))
```

Comment:
I know that the spec calls this "decryption using a FVK", but the only thing we actually use here from `fvk` is `ovk`, which is effectively random bytes. And the generated test vectors expose `ovk` directly, not the sender's `fvk`. So let's remove the `fvk` requirement:
```suggestion
    def decrypt_using_ovk(self, ovk, rseed, rho, cv, cm_star):
        ock = prf_ock_orchard(ovk, bytes(cv), bytes(cm_star.extract()), bytes(self.epk))
```